### ISSUE TITLE
-Werror=sign-conversion on `fail()`'s unsigned line number

### ISSUE THREAD

**Issue by hostilefork** [akrzemi1/Optional#33]
Hey there.  I've actually still got a C++11 project that has a dependency on this `optional.hpp`... and don't want to bump the project to C++17.  So this still seems the best canon source for the file.

In the compiler settings this has `-Werror` for all warnings as errors, and then among flags it checks is `-Wsign-conversion`.  Here `fail()` takes in an `unsigned` and passes it to `__assert( )`:

https://github.com/akrzemi1/Optional/blob/c7be0ea4b66cc9760713087ea3b5b5e54f322260/optional.hpp#L191

FWIW, the definition in <assert.h> that I have on the machine (in gcc 4.9.2-10 and clang 3.5.0-10) is:

    /* The following is not at all used here but needed for standard
       compliance.  */
    extern void __assert (const char *__assertion, const char *__file, int __line)
         __THROW __attribute__ ((__noreturn__));

It's easy enough for me to change.  But as I'm leaving notes in the file saying where it came from, it would be nice if it "ran unmodified" so that if it were pulled down from git again one wouldn't have to figure out what the error is.  *(The person doing that and getting confused would most likely just be me, later.  :-)  But it could be someone else.)*

What I'd do would be just take a signed int to fail(), due to the seemingly-prevailing wisdom that unsigned integers aren't worth it:

http://stackoverflow.com/questions/10168079/why-is-size-t-unsigned

Or it could be cast...

**Comment by akrzemi1** [comment]
Can you check if it works for you now?

**Comment by hostilefork** [comment]
That handles my case, thanks for the change!

P.S. Would you have any relevant links on specializing this version *(or a current or future version of std::optional)* in such a way that it could allow a contained type to "conspire" in order to not use any additional storage space?  I wrote an article here about it a while ago (feedback welcome):

http://blog.hostilefork.com/template-specialize-optional-or-not/

I managed to get a crude specialization working, and from the user's perspective they didn't know they were using a "new" type.  But it basically shared no code with optional.  If you had any input...I'm curious on your opinion about if it should be done at all *(if you missed that question on SO by chance)*.  If so if there is any trickery to be done to save rewriting every method in the specialization.

Thanks again!

**Comment by akrzemi1** [comment]
If I got the idea right, this could only work for a subset of types where you know that there exists a bit pattern that never represents a value.

Have you seen [compact_optional](https://github.com/akrzemi1/compact_optional)? I think it implements your idea. But with a separate interface.

**Comment by hostilefork** [comment]
@akrzemi1 Aha!  No, I had not seen that..I'm glad I asked as this is exactly the kind of thing I was hoping you would have a link to.

If you would like to mention it on that stackoverflow question (I found it, but not your library) then I will upvote it and then bounty reward it so it is up top where people see it.  :-)  Then I have to update my article too...

But I went ahead and created an issue there covering an idea regarding the difference:

https://github.com/akrzemi1/compact_optional/issues/2
